### ISSUE TITLE
Honor exactly one `prefer(x)` dependency constraint in the presence of multiple conflicting preferences

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle#4186]
When there are multiple conflicting `prefer(x)` dependency constraints present, these constraints are ignored. Instead, a version that matches at least one of these constraints should be selected, when they do not conflict with any other constraints. The same mechanism that is used to select a version from multiple candidates should be used, which currently happens to be to select the highest version.

For example, assuming versions 1.0.0, 1.1.0 and 1.2.0 are available, the following selects 1.2.0 but should select 1.1.0:

```
dependencies {
    constraints {
        implementation(group: 'org.gradle.sample', name: 'versions') {
            version {
                prefer('1.1.0')
            }
        }
        implementation(group: 'org.gradle.sample', name: 'versions') {
            version {
                prefer('1.0.0')
            }
        }
    }
    implementation 'org.gradle.sample:versions:[1.0.0,2.0.0)'
}
```

Furthermore, enabling strict conflict resolution causes this build to fail, but it should not as there are no conflicting requirements.

### Expected Behavior

A version that matches at least one preferred version should be selected when there are multiple preferred versions.

### Current Behavior

The preferred version constraints are ignored when there they conflict with each other.

**Comment by melix** [comment]
This is not specific to constraints: if multiple transitive regular dependencies have the same versions as in the example above, we select 1.2.0 too. This is a known issue, that we often refer to internally as "do best effort with ranges to select the best matching version".

**Comment by bigdaz** [comment]
This was fixed in Gradle 4.7: when multiple `prefer` version constraints are discovered, the highest one is honoured. Gradle will attempt to find the best matching version considering all input constraints,  including version ranges.

The behaviour for `failOnVersionConflict()` has not been fixed: a `prefer(x)` constraint is strongly honored in the presence of other `prefer` constraints, causing resolution to fail with `failOnVersionConflict()`. This should be addressed as part of #4181.
